Exercises I

Run python -m pip install pymysql from your terminal to install pymysql.

Create a notebook or python script named advanced_dataframes to do your work in for these exercises.

1.Run python -m pip install pymysql from your terminal to install the mysql client (any folder is fine)

2.cd into your exercises folder for this module and run echo env.py >> .gitignore

3.Create a function named get_db_url. It should accept a username, hostname, password, and database name and return an url connection string formatted like in the example at the start of this lesson.

4.Use your function to obtain a connection to the employees database.

5.Once you have successfully run a query:

a. Intentionally make a typo in the database url. What kind of error message do you see?

b. Intentionally make an error in your SQL query. What does the error message look like?

6.Read the employees and titles tables into two separate DataFrames.

7.How many rows and columns do you have in each DataFrame? Is that what you expected?

8.Display the summary statistics for each DataFrame.

9.How many unique titles are in the titles DataFrame?

10.What is the oldest date in the to_date column?

11.What is the most recent date in the to_date column?



1.Run python -m pip install pymysql from your terminal to install the mysql client (any folder is fine)

done

2.cd into your exercises folder for this module and run echo env.py >> .gitignore

done

3.Create a function named get_db_url. It should accept a username, hostname, password, and database name and return a url connection string formatted like in the example at the start of this lesson.

In [105]:
import pandas as pd
import numpy as np
import env

In [106]:
def get_db_url(database):
    database = database
    url = f'mysql+pymysql://{env.username}:{env.password}@{env.host}/{database}'
    return url


4.Use your function to obtain a connection to the employees database.

In [107]:
pd.read_sql('SELECT * FROM employees LIMIT 5 OFFSET 50', get_db_url("employees"))

emp_no  birth_date first_name  last_name gender   hire_date
0   10051  1953-07-28   Hidefumi      Caine      M  1992-10-15
1   10052  1961-02-26     Heping     Nitsch      M  1988-05-21
2   10053  1954-09-13     Sanjiv   Zschoche      F  1986-02-04
3   10054  1957-04-04     Mayumi  Schueller      M  1995-03-13
4   10055  1956-06-06     Georgy     Dredge      M  1992-04-27

5.Once you have successfully run a query:

a. Intentionally make a typo in the database url. What kind of error message do you see?
operationalerror
b. Intentionally make an error in your SQL query. What does the error message look like?
programmingerror

In [108]:
pd.read_sql('SELECT * FROM employees LIMIT 5 OFFSET 50', get_db_url("emploees"))

OperationalError: (pymysql.err.OperationalError) (1044, "Access denied for user 'pagel_2185'@'%' to database 'emploees'")
(Background on this error at: https://sqlalche.me/e/14/e3q8)

In [ ]:
pd.read_sql('SELECT * FROM employes LIMIT 5 OFFSET 50', get_db_url("employees"))

6.Read the employees and titles tables into two separate DataFrames.

In [ ]:
pd.read_sql('SELECT * FROM employees LIMIT 5 OFFSET 50', get_db_url("employees"))

In [ ]:
pd.read_sql('SELECT * FROM titles LIMIT 5 OFFSET 50', get_db_url("employees"))

7.How many rows and columns do you have in each DataFrame? Is that what you expected?

5 rows 6 columns and 5 rows 4 columns
yes that is what i was explecting

8.Display the summary statistics for each DataFrame.

In [109]:
pd.read_sql('select count(distinct title)from titles', get_db_url("employees"))

count(distinct title)
0                      7

9.How many unique titles are in the titles DataFrame?

In [110]:
pd.read_sql('select count(distinct title)from titles', get_db_url("employees"))

count(distinct title)
0                      7

10.What is the oldest date in the to_date column?

In [111]:
pd.read_sql('select min(to_date)from titles', get_db_url("employees"))

min(to_date)
0   1985-03-01

11.What is the most recent date in the to_date column?

In [112]:
pd.read_sql('''select max(to_date) as most_recent_to_date
from (select to_date
from titles
where to_date <= now())
as no_longer_here''', get_db_url("employees"))

most_recent_to_date
0          2002-08-01

Exercises II

1.Copy the users and roles DataFrames from the examples above.

2.What is the result of using a right join on the DataFrames?

3.What is the result of using an outer join on the DataFrames?

4.What happens if you drop the foreign keys from the DataFrames and try to merge them?

5.Load the mpg dataset from PyDataset.

6.Output and read the documentation for the mpg dataset.

7.How many rows and columns are in the dataset?

8.Check out your column names and perform any cleanup you may want on them.

9.Display the summary statistics for the dataset.

10.How many different manufacturers are there?

11.How many different models are there?

12.Create a column named mileage_difference like you did in the DataFrames exercises; this column should contain the difference between highway and city mileage for each car.

13.Create a column named average_mileage like you did in the DataFrames exercises; this is the mean of the city and highway mileage.

14.Create a new column on the mpg dataset named is_automatic that holds boolean values denoting whether the car has an automatic transmission.

15.Using the mpg dataset, find out which manufacturer has the best miles per gallon on average?

16.Do automatic or manual cars have better miles per gallon?



1.Copy the users and roles DataFrames from the examples above.

In [113]:
# Create the users DataFrame.

users = pd.DataFrame({
    'id': [1, 2, 3, 4, 5, 6],
    'name': ['bob', 'joe', 'sally', 'adam', 'jane', 'mike'],
    'role_id': [1, 2, 3, 3, np.nan, np.nan]
})
users


id   name  role_id
0   1    bob      1.0
1   2    joe      2.0
2   3  sally      3.0
3   4   adam      3.0
4   5   jane      NaN
5   6   mike      NaN

In [114]:
# Create the roles DataFrame

roles = pd.DataFrame({
    'id': [1, 2, 3, 4],
    'name': ['admin', 'author', 'reviewer', 'commenter']
})
roles


id       name
0   1      admin
1   2     author
2   3   reviewer
3   4  commenter

2.What is the result of using a right join on the DataFrames?

In [115]:
right_merge = roles.merge(users, on='id', how='right')
right_merge


id     name_x name_y  role_id
0   1      admin    bob      1.0
1   2     author    joe      2.0
2   3   reviewer  sally      3.0
3   4  commenter   adam      3.0
4   5        NaN   jane      NaN
5   6        NaN   mike      NaN

3.What is the result of using an outer join on the DataFrames?

In [116]:
outer_merge = users.merge(roles, on='id', how='outer')
outer_merge

id name_x  role_id     name_y
0   1    bob      1.0      admin
1   2    joe      2.0     author
2   3  sally      3.0   reviewer
3   4   adam      3.0  commenter
4   5   jane      NaN        NaN
5   6   mike      NaN        NaN

4.What happens if you drop the foreign keys from the DataFrames and try to merge them?
doesnt show anything


In [117]:
foreign_merge = roles.merge(users, on='name', how='inner')
foreign_merge

Empty DataFrame
Columns: [id_x, name, id_y, role_id]
Index: []

5.Load the mpg dataset from PyDataset.

In [118]:
from pydataset import data
mpg_df = data('mpg')
mpg_df

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

6.Output and read the documentation for the mpg dataset.

In [119]:
mpg_df

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

7.How many rows and columns are in the dataset?

234 rows x 11 columns

8.Check out your column names and perform any cleanup you may want on them.

In [120]:
mpg_df.sort_values("hwy")

manufacturer                model  displ  year  cyl       trans drv  cty  \
55         dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4    9   
66         dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4    9   
127         jeep   grand cherokee 4wd    4.7  2008    8    auto(l5)   4    9   
70         dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4    9   
60         dodge          durango 4wd    4.7  2008    8    auto(l5)   4    9   
..           ...                  ...    ...   ...  ...         ...  ..  ...   
107        honda                civic    1.8  2008    4    auto(l5)   f   24   
197       toyota              corolla    1.8  2008    4  manual(m5)   f   28   
223   volkswagen           new beetle    1.9  1999    4    auto(l4)   f   29   
213   volkswagen                jetta    1.9  1999    4  manual(m5)   f   33   
222   volkswagen           new beetle    1.9  1999    4  manual(m5)   f   35   

     hwy fl       class  
55    12  e      pickup  
66    12  e      pickup  
127   12  e         suv  
70    12  e      pickup  
60    12  e         suv  
..   ... ..         ...  
107   36  c  subcompact  
197   37  r     compact  
223   41  d  subcompact  
213   44  d     compact  
222   44  d  subcompact  

[234 rows x 11 columns]


9.Display the summary statistics for the dataset.

In [121]:
mpg_df.info()
mpg_df.describe()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 234 entries, 1 to 234
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   manufacturer  234 non-null    object 
 1   model         234 non-null    object 
 2   displ         234 non-null    float64
 3   year          234 non-null    int64  
 4   cyl           234 non-null    int64  
 5   trans         234 non-null    object 
 6   drv           234 non-null    object 
 7   cty           234 non-null    int64  
 8   hwy           234 non-null    int64  
 9   fl            234 non-null    object 
 10  class         234 non-null    object 
dtypes: float64(1), int64(4), object(6)
memory usage: 21.9+ KB


displ         year         cyl         cty         hwy
count  234.000000   234.000000  234.000000  234.000000  234.000000
mean     3.471795  2003.500000    5.888889   16.858974   23.440171
std      1.291959     4.509646    1.611534    4.255946    5.954643
min      1.600000  1999.000000    4.000000    9.000000   12.000000
25%      2.400000  1999.000000    4.000000   14.000000   18.000000
50%      3.300000  2003.500000    6.000000   17.000000   24.000000
75%      4.600000  2008.000000    8.000000   19.000000   27.000000
max      7.000000  2008.000000    8.000000   35.000000   44.000000


10.How many different manufacturers are there? 15

In [122]:
mpg_df.groupby('manufacturer').describe()

displ                                                       year  \
             count      mean           std  min    25%  50%    75%  max count   
manufacturer                                                                    
audi          18.0  2.544444  6.732032e-01  1.8  2.000  2.8  3.025  4.2  18.0   
chevrolet     19.0  5.063158  1.370406e+00  2.4  4.450  5.3  5.850  7.0  19.0   
dodge         37.0  4.378378  8.679910e-01  2.4  3.800  4.7  4.700  5.9  37.0   
ford          25.0  4.536000  5.407402e-01  3.8  4.000  4.6  4.600  5.4  25.0   
honda          9.0  1.711111  1.452966e-01  1.6  1.600  1.6  1.800  2.0   9.0   
hyundai       14.0  2.428571  3.646464e-01  2.0  2.100  2.4  2.650  3.3  14.0   
jeep           8.0  4.575000  1.015241e+00  3.0  3.925  4.7  4.950  6.1   8.0   
land rover     4.0  4.300000  2.581989e-01  4.0  4.150  4.3  4.450  4.6   4.0   
lincoln        3.0  5.400000  1.087792e-15  5.4  5.400  5.4  5.400  5.4   3.0   
mercury        4.0  4.400000  4.898979e-01  4.0  4.000  4.3  4.700  5.0   4.0   
nissan        13.0  3.269231  8.644696e-01  2.4  2.500  3.3  3.500  5.6  13.0   
pontiac        5.0  3.960000  8.080842e-01  3.1  3.800  3.8  3.800  5.3   5.0   
subaru        14.0  2.457143  1.089410e-01  2.2  2.500  2.5  2.500  2.5  14.0   
toyota        34.0  2.952941  9.313513e-01  1.8  2.250  2.7  3.400  5.7  34.0   
volkswagen    27.0  2.255556  4.431820e-01  1.8  2.000  2.0  2.500  3.6  27.0   

                           ...    cty         hwy                             \
                     mean  ...    75%   max count       mean       std   min   
manufacturer               ...                                                 
audi          2003.500000  ...  18.75  21.0  18.0  26.444444  2.175322  23.0   
chevrolet     2004.684211  ...  16.50  22.0  19.0  21.894737  5.108759  14.0   
dodge         2004.108108  ...  15.00  18.0  37.0  17.945946  3.574182  12.0   
ford          2002.600000  ...  15.00  18.0  25.0  19.360000  3.327662  15.0   
honda         2003.000000  ...  25.00  28.0   9.0  32.555556  2.554952  29.0   
hyundai       2004.142857  ...  19.75  21.0  14.0  26.857143  2.178819  24.0   
jeep          2005.750000  ...  15.00  17.0   8.0  17.625000  3.248626  12.0   
land rover    2003.500000  ...  12.00  12.0   4.0  16.500000  1.732051  15.0   
lincoln       2002.000000  ...  11.50  12.0   3.0  17.000000  1.000000  16.0   
mercury       2003.500000  ...  13.25  14.0   4.0  18.000000  1.154701  17.0   
nissan        2003.846154  ...  19.00  23.0  13.0  24.615385  5.091471  17.0   
pontiac       2002.600000  ...  18.00  18.0   5.0  26.400000  1.140175  25.0   
subaru        2004.142857  ...  20.00  21.0  14.0  25.571429  1.157868  23.0   
toyota        2002.705882  ...  21.00  28.0  34.0  24.911765  6.166221  15.0   
volkswagen    2002.666667  ...  21.00  35.0  27.0  29.222222  5.315676  23.0   

                                        
                25%   50%    75%   max  
manufacturer                            
audi          25.00  26.0  27.75  31.0  
chevrolet     17.00  23.0  26.00  30.0  
dodge         16.00  17.0  21.00  24.0  
ford          17.00  18.0  22.00  26.0  
honda         32.00  32.0  34.00  36.0  
hyundai       26.00  26.5  28.00  31.0  
jeep          16.25  18.5  19.25  22.0  
land rover    15.00  16.5  18.00  18.0  
lincoln       16.50  17.0  17.50  18.0  
mercury       17.00  18.0  19.00  19.0  
nissan        20.00  26.0  27.00  32.0  
pontiac       26.00  26.0  27.00  28.0  
subaru        25.00  26.0  26.00  27.0  
toyota        20.00  26.0  29.75  37.0  
volkswagen    26.00  29.0  29.00  44.0  

[15 rows x 40 columns]


11.How many different models are there? 38

In [123]:
mpg_df.groupby('model').describe()

displ                                                   \
                       count      mean           std  min    25%   50%    75%   
model                                                                           
4runner 4wd              6.0  3.483333  7.730890e-01  2.7  2.875  3.40  3.850   
a4                       7.0  2.328571  5.498918e-01  1.8  1.900  2.00  2.800   
a4 quattro               8.0  2.425000  5.775564e-01  1.8  1.950  2.40  2.875   
a6 quattro               3.0  3.366667  7.371115e-01  2.8  2.950  3.10  3.650   
altima                   6.0  2.800000  5.440588e-01  2.4  2.425  2.50  3.250   
c1500 suburban 2wd       5.0  5.520000  3.193744e-01  5.3  5.300  5.30  5.700   
camry                    7.0  2.671429  4.990467e-01  2.2  2.300  2.40  3.000   
camry solara             7.0  2.642857  4.466809e-01  2.2  2.300  2.40  3.000   
caravan 2wd             11.0  3.390909  4.526689e-01  2.4  3.300  3.30  3.800   
civic                    9.0  1.711111  1.452966e-01  1.6  1.600  1.60  1.800   
corolla                  5.0  1.800000  0.000000e+00  1.8  1.800  1.80  1.800   
corvette                 5.0  6.160000  5.319774e-01  5.7  5.700  6.20  6.200   
dakota pickup 4wd        9.0  4.411111  6.153138e-01  3.7  3.900  4.70  4.700   
durango 4wd              7.0  4.971429  6.848705e-01  3.9  4.700  4.70  5.450   
expedition 2wd           3.0  5.133333  4.618802e-01  4.6  5.000  5.40  5.400   
explorer 4wd             6.0  4.266667  4.320494e-01  4.0  4.000  4.00  4.450   
f150 pickup 4wd          7.0  4.714286  5.014265e-01  4.2  4.400  4.60  5.000   
forester awd             6.0  2.500000  0.000000e+00  2.5  2.500  2.50  2.500   
grand cherokee 4wd       8.0  4.575000  1.015241e+00  3.0  3.925  4.70  4.950   
grand prix               5.0  3.960000  8.080842e-01  3.1  3.800  3.80  3.800   
gti                      5.0  2.160000  3.577709e-01  2.0  2.000  2.00  2.000   
impreza awd              8.0  2.425000  1.388730e-01  2.2  2.425  2.50  2.500   
jetta                    9.0  2.277778  3.700601e-01  1.9  2.000  2.00  2.500   
k1500 tahoe 4wd          4.0  5.700000  5.656854e-01  5.3  5.300  5.50  5.900   
land cruiser wagon 4wd   2.0  5.200000  7.071068e-01  4.7  4.950  5.20  5.450   
malibu                   5.0  3.000000  5.787918e-01  2.4  2.400  3.10  3.500   
maxima                   3.0  3.166667  2.886751e-01  3.0  3.000  3.00  3.250   
mountaineer 4wd          4.0  4.400000  4.898979e-01  4.0  4.000  4.30  4.700   
mustang                  9.0  4.377778  5.238745e-01  3.8  4.000  4.60  4.600   
navigator 2wd            3.0  5.400000  1.087792e-15  5.4  5.400  5.40  5.400   
new beetle               6.0  2.133333  2.875181e-01  1.9  1.925  2.00  2.375   
passat                   7.0  2.400000  6.831301e-01  1.8  1.900  2.00  2.800   
pathfinder 4wd           4.0  4.050000  1.084743e+00  3.3  3.300  3.65  4.400   
ram 1500 pickup 4wd     10.0  5.020000  4.613988e-01  4.7  4.700  4.70  5.200   
range rover              4.0  4.300000  2.581989e-01  4.0  4.150  4.30  4.450   
sonata                   7.0  2.557143  3.309438e-01  2.4  2.400  2.40  2.500   
tiburon                  7.0  2.300000  3.741657e-01  2.0  2.000  2.00  2.700   
toyota tacoma 4wd        7.0  3.271429  5.879747e-01  2.7  2.700  3.40  3.700   

                             year               ...    cty         hwy  \
                        max count         mean  ...    75%   max count   
model                                           ...                      
4runner 4wd             4.7   6.0  2002.000000  ...  15.75  16.0   6.0   
a4                      3.1   7.0  2002.857143  ...  20.50  21.0   7.0   
a4 quattro              3.1   8.0  2003.500000  ...  18.25  20.0   8.0   
a6 quattro              4.2   3.0  2005.000000  ...  16.50  17.0   3.0   
altima                  3.5   6.0  2005.000000  ...  22.50  23.0   6.0   
c1500 suburban 2wd      6.0   5.0  2006.200000  ...  14.00  14.0   5.0   
camry                   3.5   7.0  


12.Create a column named mileage_difference like you did in the DataFrames exercises; this column should contain the difference between highway and city mileage for each car.

In [137]:
mpg_df['mileage_difference'] = abs(mpg_df.hwy - mpg_df.cty)


13.Create a column named average_mileage like you did in the DataFrames exercises; this is the mean of the city and highway mileage.

In [136]:
mpg_df['average_mileage'] = round((mpg_df.hwy + mpg_df.cty) / 2, 1)


14.Create a new column on the mpg dataset named is_automatic that holds boolean values denoting whether the car has an automatic transmission.

In [131]:
mpg_df['is_automatic'] = mpg_df["trans"].str.contains("auto")


15.Using the mpg dataset, find out which manufacturer has the best miles per gallon on average?

In [147]:
mpg_by_manufacturer = mpg_df.groupby('manufacturer')['average_mileage'].mean()
best_manufacturer = mpg_by_manufacturer.idxmax()
best_manufacturer

'honda'


16.Do automatic or manual cars have better miles per gallon?

In [149]:
mpg_by_trans = mpg_df.groupby('is_automatic')['average_mileage'].mean()
mpg_by_trans

is_automatic
False    22.227273
True     19.130573
Name: average_mileage, dtype: float64

Exercises III

1.Use your get_db_url function to help you explore the data from the chipotle database.

2.What is the total price for each order?

3.What are the most popular 3 items?

4.Which item has produced the most revenue?

5.Join the employees and titles DataFrames together.

6.For each title, find the hire date of the employee that was hired most recently with that title.

7.Write the code necessary to create a cross tabulation of the number of titles by department. (Hint: this will involve a combination of SQL code to pull the necessary data and python/pandas code to perform the manipulations.)